src/libsyntax/parse/parser.rs

```diff
@@ -2813,16 +2813,25 @@ impl<'a> Parser<'a> {
 
 
             let rhs = try!(match op.fixity() {
-                Fixity::Right => self.with_res(restrictions, |this|{
-                    this.parse_assoc_expr_with(op.precedence(), LhsExpr::NotYetParsed)
+                Fixity::Right => self.with_res(
+                    restrictions & !Restrictions::RESTRICTION_STMT_EXPR,
```

Comment:
This can be written as `restrictions - Restrictions::RESTRICTION_STMT_EXPR`, AFAIK.